### PR TITLE
bpo-29942: Fix the use of recursion in itertools.chain.from_iterable.

### PR DESCRIPTION
**Pull request by Yhg1s**
Fix the use of recursion in `itertools.chain.from_iterable`. Using recursion is unnecessary, and can easily cause stack overflows, especially when building in low optimization modes or with `Py_DEBUG` enabled.

### CODE REVIEW

**Comment by serhiy-storchaka** [comment]
Please open an issue on the bug tracker, add an entry in `Misc/NEWS` at the start of the "Library" section (don't forget to add "Patch by <your name>."), and add your name in `Misc/ACK`.

**Comment by serhiy-storchaka** [comment]
The patch itself LGTM.

**Comment by Yhg1s** [comment]
bugs.python.org issue is http://bugs.python.org/issue29942. Misc/NEWS entry added. I'm already in Misc/ACKS, and considering I've been a core dev for many years I don't think I need separate credits for the patch.

**Comment by serhiy-storchaka** [comment]
Oh, sorry. Your GitHab name was not known to me.

File: Lib/test/test_itertools.py [hunk 1]
```diff
@@ -1976,6 +1976,14 @@ def gen2(x):
         self.assertRaises(AssertionError, list, cycle(gen1()))
         self.assertEqual(hist, [0,1])
 
+    def test_deep_recursion(self):
+        # Make sure itertools.chain doesn't run into recursion limits when
+        # dealing with long chains of empty iterables. Even with a high
+        # number this would probably only fail in Py_DEBUG mode.
+        it = chain.from_iterable(() for unused in range(10000000))
+        with self.assertRaises(StopIteration):
+          next(it)
```
Comment (serhiy-storchaka) [inline Lib/test/test_itertools.py]: Two-space indentation?

File: Lib/test/test_itertools.py [hunk 2]
```diff
@@ -1976,6 +1976,14 @@ def gen2(x):
         self.assertRaises(AssertionError, list, cycle(gen1()))
         self.assertEqual(hist, [0,1])
 
+    def test_deep_recursion(self):
```
Comment (serhiy-storchaka) [inline Lib/test/test_itertools.py]: Now this is not a recursion. The recursion from user side is `chain.from_iterable(chain.from_iterable(...))`. Maybe use other name for the test?

Comment (Yhg1s) [reply Lib/test/test_itertools.py]: Name adjusted.

File: Lib/test/test_itertools.py (hunk 1 above)
Comment (Yhg1s) [reply Lib/test/test_itertools.py]: Fixed.

**Review by serhiy-storchaka — APPROVED** [review]

**Review by rhettinger — APPROVED** [review]

**Pull request merged**